### ISSUE TITLE
security/acme-client: certificate status not correctly updated

### ISSUE THREAD

**Issue by lrosenman** [opnsense/plugins#1307]
I have added my FW's domain, etc to the Let's Encrypt feature, and the certificate is fine, but the status field indicates validation failed. 

<img width="1853" alt="Screen Shot 2019-04-22 at 2 10 47 PM" src="https://user-images.githubusercontent.com/2868525/56519736-72b32e80-6508-11e9-83df-e9cfd9038114.png">

What else do you need?

**Comment by lrosenman** [comment]
<img width="757" alt="Screen Shot 2019-04-22 at 2 11 49 PM" src="https://user-images.githubusercontent.com/2868525/56519815-9d04ec00-6508-11e9-8c57-0dc65125927c.png">

**Comment by lrosenman** [comment]
I **INTENTIONALLY** did not show my key, but it's there.

**Comment by lrosenman** [comment]
@fichtner or anyone else, can y'all look at this?

**Comment by fichtner** [comment]
@fraenki will probably take a look, but please give him time. I am not using LE.

**Comment by fraenki** [comment]
@lrosenman Please have a look at the log file: Services -> Let's Encrypt -> Log File. The log is bloated and looks horrible in the GUI, but it usually reveals what the problem is. :)

**Comment by lrosenman** [comment]
It does *NOT*.
[Fri Apr 12 19:19:26 CDT 2019]	_on_issue_success
[Fri Apr 12 19:19:26 CDT 2019]	Installing full chain to:/var/etc/acme-client/certs/5ca501a083ac17.11218680/fullchain.pem
[Fri Apr 12 19:19:26 CDT 2019]	Installing key to:/var/etc/acme-client/keys/5ca501a083ac17.11218680/private.key
[Fri Apr 12 19:19:26 CDT 2019]	Installing CA to:/var/etc/acme-client/certs/5ca501a083ac17.11218680/chain.pem
[Fri Apr 12 19:19:26 CDT 2019]	Installing cert to:/var/etc/acme-client/certs/5ca501a083ac17.11218680/cert.pem
[Fri Apr 12 19:19:26 CDT 2019]	And the full chain certs is there: /var/etc/acme-client/home/home-fw.lerctr.org/fullchain.cer
[Fri Apr 12 19:19:26 CDT 2019]	The intermediate CA cert is in /var/etc/acme-client/home/home-fw.lerctr.org/ca.cer
[Fri Apr 12 19:19:26 CDT 2019]	ret='0'
[Fri Apr 12 19:19:26 CDT 2019]	_CURL='curl -L --silent --dump-header /var/etc/acme-client/home/http.header -g '
[Fri Apr 12 19:19:26 CDT 2019]	timeout=
[Fri Apr 12 19:19:26 CDT 2019]	url='https://acme-v01.api.letsencrypt.org/acme/issuer-cert'
[Fri Apr 12 19:19:26 CDT 2019]	GET
[Fri Apr 12 19:19:26 CDT 2019]	_link_issuer_retry='0'
[Fri Apr 12 19:19:26 CDT 2019]	Le_LinkIssuer='https://acme-v01.api.letsencrypt.org/acme/issuer-cert'
[Fri Apr 12 19:19:26 CDT 2019]	Your cert key is in /var/etc/acme-client/home/home-fw.lerctr.org/home-fw.lerctr.org.key
[Fri Apr 12 19:19:26 CDT 2019]	Your cert is in /var/etc/acme-client/home/home-fw.lerctr.org/home-fw.lerctr.org.cer
[Fri Apr 12 19:19:26 CDT 2019]	Cert success.
[Fri Apr 12 19:19:26 CDT 2019]	Le_LinkCert='https://acme-v01.api.letsencrypt.org/acme/cert/0384492db2c653f066f132e77a7bb7036f93'
[Fri Apr 12 19:19:26 CDT 2019]	code='201'
[Fri Apr 12 19:19:26 CDT 2019]	_ret='0'
[Fri Apr 12 19:19:23 CDT 2019]	_CURL='curl -L --silent --dump-header /var/etc/acme-client/home/http.header -g '
[Fri Apr 12 19:19:23 CDT 2019]	_post_url='https://acme-v01.api.letsencrypt.org/acme/new-cert'
[Fri Apr 12 19:19:23 CDT 2019]	POST
[Fri Apr 12 19:19:23 CDT 2019]	payload='{"resource": "new-cert", "csr": "MIIEsDCCApgCAQAwHTEbMBkGA1UEAwwSaG9tZS1mdy5sZXJjdHIub3JnMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAt24E9YsDK6k94VhKPIL8b-xH2YgxM5HqYaoIlxztxSRPXnYXJjpAqN6LTuhwM49hi4E-HZyGdNtIiFeXET5CC8z8D-usRD3Dy2L1lvb6azQrZXFfd1XASnc0WW_lJByM6DmjpRYgnrAzx2nnbbvTQ1Cskie-5hI2X1Xqts5T_UN7zsVnv7wbpiT76A2OwvsMIwivPsFDwR6ELEORw4cotltGW5-JXzAYH8yVh4yaaBkFiHnhiFgva-amJc_B_KxBftRkjLW3FCDiZJTqMr2VbPQyhQHi1U71vGu-oBpPBY_jlfJnrdWac6TXm7RdtSM8aq7WzumbsXAr7RzFcz0nIY89IAs39HaIrZNTWQqiCADw1raReZbuyYU8XYgzJcgPanOmOqEwGk-RZsoqI_du4_c8kIRGCsMmN8QHqKGkvPRj3K-GzcGiwYzl2e7aa4gzBEpRxz31sLO7hceTUWGf1_yWKXI8y2d43oGi5TOmZq5ttgAqgzqZ69fXY91gs1WzBLy3AUBPDzG2lMBxTNhO6PtMnnMO6qk3KBFvEXiz1lOGQusgKuefFaJiQWXH9clLDVelVY-YKK2Lmi-zHqElC8ga28aTfGItqs1eKr6BS08-RDDk1lwJCeGNYugJyzq5joUN3PJJv3aV6UgwlFh65j5yA7PbxNLkBjbQFbezmHsCAwEAAaBOMEwGCSqGSIb3DQEJDjE_MD0wCwYDVR0PBAQDAgXgMC4GA1UdEQQnMCWCEmhvbWUtZncubGVyY3RyLm9yZ4IPaG9tZS5sZXJjdHIub3JnMA0GCSqGSIb3DQEBCwUAA4ICAQAP6l_hazFt6fv-4jySBc7XZ_Xqj1NFNB0By-F7xbFCGIpAO7EFlWZ2VQme-RrqtNeUxkiC9S1fi1QZDgrKE3qmk7BK87CW8p04dHn5E7vCENlXxQ13w7oP5KkTs6hbRi6XDMJTx_2EYTIpLseY3HIBRz7PK6BtEk6R-2_HuDEYiV1OqwIVt5Wd3KQR908mBq3mRPErhhjA_RbKRn5ae6fB35lFoUryggkcLEPdphT1x5vPAiGKilm6JgFNy0UTW6NvI4TiEs5QMlmI8n3pmR8ZjZlqnWjr1JFOBHObwwZRIJTqObRHCWmU8zATv8CYOGpMm4Z5e28aMp31qZr57vev9TtsnPedBc7ZZAF6a--ktZ6SbVBywzgB02Sv7sTNFMeD2hvoJP2GBgU7_6zxjbvpqe04J8JXRStAywb0_46a9jDB0k-7SgI544PatXBtNfTkj4J4QgyQBK2-jux3vnSD-qaFwOB9fly50WQFrwAXPLKFmTZfPRgXcWSuTFRiTHZziOtAhd-Im6Qn2Dsoh_dgOCW5hQxON_RHT5DvGsxVGqdfkY9__ooDjK8d3GqlVQmfynevBYytE2b-uJlTMBwFPPzJ9-ZyXb_5gtvERkrgWzCHOFshxg801EDF0Pt-rc1WGgPXigqCJu4YUZLg2olT4BLXRSGQs3V6Ru62HuGRUQ"}'
[Fri Apr 12 19:19:23 CDT 2019]	url='https://acme-v01.api.letsencrypt.org/acme/new-cert'
[Fri Apr 12 19:19:23 CDT 2019]	j='27'
[Fri Apr 12 19:19:23 CDT 2019]	i='2'
[Fri Apr 12 19:19:23 CDT 2019]	Verify finished, start to sign.
[Fri Apr 12 19:19:23 CDT 2019]	skip dns.
[Fri Apr 12 19:19:23 CDT 2019]	vlist='home-fw.lerctr.org#verified_ok#https://acme-v01.api.letsencrypt.org/acme/challenge/_htoKG9bP67f88rvuWhYZ38txJXuMhB_FHP_41J9Esg/14362201100#dns-01#dns_nsupdate,home.lerctr.org#verified_ok#https://acme-v01.api.letsencrypt.org/acme/challenge/cpaYRIrQ9fWtKWNmtVWVjuuG79fxKI9zm5N7exv8aS8/14400602746#dns-01#dns_nsupdate,'
[Fri Apr 12 19:19:23 CDT 2019]	dnsadded
[Fri Apr 12 19:19:23 CDT 2019]	_clearupdns
[Fri Apr 12 19:19:23 CDT 2019]	No need to restore nginx, skip.
[Fri Apr 12 19:19:23 CDT 2019]	pid
[Fri Apr 12 19:19:23 CDT 2019]	home.lerctr.org is already verified, skip dns-01.
[Fri Apr 12 19:19:23 CDT 2019]	home-fw.lerctr.org is already verified, skip dns-01.
[Fri Apr 12 19:19:23 CDT 2019]	ok, let's start to verify
[Fri Apr 12 19:19:23 CDT 2019]	home.lerctr.org is already verified, skip dns-01.
[Fri Apr 12 19:19:23 CDT 2019]	d='home.lerctr.org'
[Fri Apr 12 19:19:23 CDT 2019]	home-fw.lerctr.org is already verified, skip dns-01.
[Fri Apr 12 19:19:23 CDT 2019]	d='home-fw.lerctr.org'
[Fri Apr 12 19:19:23 CDT 2019]	vlist='home-fw.lerctr.org#verified_ok#https://acme-v01.api.letsencrypt.org/acme/challenge/_htoKG9bP67f88rvuWhYZ38txJXuMhB_FHP_41J9Esg/14362201100#dns-01#dns_nsupdate,home.lerctr.org#verified_ok#https://acme-v01.api.letsencrypt.org/acme/challenge/cpaYRIrQ9fWtKWNmtVWVjuuG79fxKI9zm5N7exv8aS8/14400602746#dns-01#dns_nsupdate,'
[Fri Apr 12 19:19:23 CDT 2019]	d
[Fri Apr 12 19:19:23 CDT 2019]	dvlist='home.lerctr.org#verified_ok#https://acme-v01.api.letsencrypt.org/acme/challenge/cpaYRIrQ9fWtKWNmtVWVjuuG79fxKI9zm5N7exv8aS8/14400602746#dns-01#dns_nsupdate'
[Fri Apr 12 19:19:23 CDT 2019]	keyauthorization='verified_ok'
[Fri Apr 12 19:19:23 CDT 2019]	home.lerctr.org is already verified.
[Fri Apr 12 19:19:23 CDT 2019]	keyauthorization='ewXX5yMrDG7Un8LOFyseeKAu5_Gj_xCWg5ufQrhqz6Q.adLS3Ytv3QdAYEZGt0yL0hERLr5KZw7_cXTVPP3oqtw'
[Fri Apr 12 19:19:23 CDT 2019]	uri='https://acme-v01.api.letsencrypt.org/acme/challenge/cpaYRIrQ9fWtKWNmtVWVjuuG79fxKI9zm5N7exv8aS8/14400602746'
[Fri Apr 12 19:19:23 CDT 2019]	token='ewXX5yMrDG7Un8LOFyseeKAu5_Gj_xCWg5ufQrhqz6Q'
[Fri Apr 12 19:19:23 CDT 2019]	entry='"type":"dns-01","status":"valid","uri":"https://acme-v01.api.letsencrypt.org/acme/challenge/cpaYRIrQ9fWtKWNmtVWVjuuG79fxKI9zm5N7exv8aS8/14400602746","token":"ewXX5yMrDG7Un8LOFyseeKAu5_Gj_xCWg5ufQrhqz6Q","validationRecord":[{"hostname":"home.lerctr.org"'
OPNsense (c) 2014-2019 Deciso B.V.

and the certificate is *FINE*. 

What else would you suggest?
'

**Comment by fraenki** [comment]
@lrosenman That's strange. Can you check the system log for additional errors related to AcmeClient? Some users expierence a "node no longer exists" error. If you see this or a similar error, then this might explain why the certificate status wasn't updated.

I'll add additional log messages to the system log with the next release, hopefully this will help to figure out what went wrong.

**Comment by lrosenman** [comment]
I *DID* see those messages *WHEN I FIRST INSTALLED*.  I can't seem to get them to come out again, NOR can I get it to even *TRY* to run the validation .

I can force issue certificates all day long, but my concern is the auto-renew.

The log from the first install is *LONG* gone.

**Comment by fraenki** [comment]
> I can force issue certificates all day long, but my concern is the auto-renew.

I've just checked one of my production setups: renewal seems to be working. I assume that only the status is not updated correctly.
But I ask you to wait for auto-renew and post your logs, if it fails to renew the certificate.

**Comment by imidoriya** [comment]
I also have the "**validation failed**" message in the **Last Acme Status** for let's encrypt even though it appears to be a success.  I'm using HTTP01 though.  I turned on debug logging using the staging.  Here is my output.. redacted.  It issues for the root domain and a couple alias hosts, which all properly resolve.  Also tried it with just the two hosts.

> [Wed May 8 21:51:32 UTC 2019]	**'' does not contain 'dns'**
> [Wed May 8 21:51:32 UTC 2019]	_on_issue_success
> [Wed May 8 21:51:32 UTC 2019]	Installing full chain to:/var/etc/acme-client/certs/...redacted.../fullchain.pem
> [Wed May 8 21:51:32 UTC 2019]	Installing key to:/var/etc/acme-client/keys/...redacted.../private.key
> [Wed May 8 21:51:32 UTC 2019]	Installing CA to:/var/etc/acme-client/certs/...redacted.../chain.pem
> [Wed May 8 21:51:32 UTC 2019]	Installing cert to:/var/etc/acme-client/certs/...redacted.../cert.pem
> [Wed May 8 21:51:32 UTC 2019]	And the full chain certs is there: /var/etc/acme-client/home/...redacted...
> [Wed May 8 21:51:32 UTC 2019]	The intermediate CA cert is in /var/etc/acme-client/home/...redacted...
> [Wed May 8 21:51:32 UTC 2019]	ret='0'
> [Wed May 8 21:51:32 UTC 2019]	_CURL='curl -L --silent --dump-header /var/etc/acme-client/home/http.header -g '
> [Wed May 8 21:51:32 UTC 2019]	timeout=
> [Wed May 8 21:51:32 UTC 2019]	url='https://acme-staging.api.letsencrypt.org/acme/issuer-cert'
> [Wed May 8 21:51:32 UTC 2019]	GET
> [Wed May 8 21:51:32 UTC 2019]	_link_issuer_retry='0'
> [Wed May 8 21:51:32 UTC 2019]	Le_LinkIssuer='https://acme-staging.api.letsencrypt.org/acme/issuer-cert'
> [Wed May 8 21:51:32 UTC 2019]	Your cert key is in /var/etc/acme-client/home/...redacted...
> [Wed May 8 21:51:32 UTC 2019]	Your cert is in /var/etc/acme-client/home/...redacted...
> [Wed May 8 21:51:32 UTC 2019]	Cert success.
> [Wed May 8 21:51:32 UTC 2019]	Le_LinkCert='https://acme-staging.api.letsencrypt.org/acme/cert/...redacted...'
> [Wed May 8 21:51:32 UTC 2019]	code='201'
> [Wed May 8 21:51:32 UTC 2019]	response='...redacted...'
> Date: Wed, 08 May 2019 21:51:32	GMT

**Comment by devNan0** [comment]
I have the same problem and added your WIP PR.

Log says:
```
May  25 00:35:00 router config[43136]:  /usr/local/opnsense/scripts/OPNsense/AcmeClient/certhelper.php:  AcmeClient: failed to set status '' for cert  XXXXXXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
```

**Comment by lrosenman** [comment]
link to the log from the auto-renew is in #1366.

**Comment by fraenki** [comment]
This will finally be fixed in the upcoming release 1.24 of os-acme-client.
